File: JavaScript/String/Ellipsis.js
function trimSpecials(text, chars) {
    var result = '';
    for (var i = 0; i < text.length; i++) {
        var isNormal = true;
        for (var j = 0; j < text.length; j++) {
            if (text[i] == chars[j]) {
                isNormal = false;
                break;
            }
        }
        if (isNormal) result += text[i];
    }
    return result;
}

function ellipsis(text, length, ellipsis) {
    ellipsis = typeof ellipsis !== 'undefined' ? ellipsis : '...';
    if (text.length <= length) return text;
    if (length <= ellipsis.length) return ellipsis.substr(0, length);
    text = text.substr(0, length - ellipsis.length);
    var position = text.lastIndexOf(' ');
    position = position < 0 ? 0 : position;
    return trimSpecials(text.substr(0, position), ['.', ',', ';', ':', '!', '?'])  + ellipsis;
}

var texto = "Este é um teste de texto longo que precisa ser truncado com reticências, sem cortar a palavra.";
for (var i = 0; i < texto.length + 1; ++i) console.log(i + ' : ' + ellipsis(texto, i));

//https://pt.stackoverflow.com/q/155204/101

File: CSharp/String/Ellipsis.cs
using static System.Console;

public class Program {
	public static void Main(string[] args) {
		string texto = "Este é um teste de texto longo que precisa ser truncado com reticências, sem cortar a palavra.";
		for (int i = 0; i < texto.Length + 1; ++i) WriteLine($"{i} : {texto.Ellipsis(i)}");
	}
}

public static class StringExt {
	static public string Ellipsis(this string text, int length, string ellipsis = "...") {
		if (text.Length <= length) return text;
		if (length <= ellipsis.Length) return ellipsis.Substring(0, length);
		text = text.Substring(0, length - ellipsis.Length);
		var position = text.LastIndexOf(' ');
		position = position < 0 ? 0 : position;
		return text.Substring(0, position).Trim(new[] { '.', ',', ';', ':', '!', '?' }) + ellipsis;
	}
}

//https://pt.stackoverflow.com/q/155204/101
